### PR TITLE
[flutter_tools] Test that DAP process terminates at the end of a session

### PR DESCRIPTION
**Pull request by DanTup**
The debug adapter should terminate itself after terminating the debugee, but previously wasn't (and was relying on the editor to do it after getting the Terminate event).

The issue is fixed in the base debug adapter by https://dart-review.googlesource.com/c/sdk/+/252660/ but this updates a test to ensure that's working correctly.

This test won't pass until https://dart-review.googlesource.com/c/sdk/+/252660/ is released in a DDS release and rolled into Flutter. I'm raising this as a draft while it's all fresh in my mind, and will mark for-review once the DDS update is ready.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by DanTup** [comment]
@christopherfujino I don't think this failure is related to the change:

https://ci.chromium.org/ui/p/flutter/builders/try/Windows%20build_tests_1_3/25422/overview

```
[   +5 ms] Gradle task assembleRelease failed with exit code 1
[        ] 
           #0      throwToolExit (package:flutter_tools/src/base/common.dart:10:3)
           #1      AndroidGradleBuilder.buildGradleApp (package:flutter_tools/src/android/gradle.dart:408:9)
           <asynchronous suspension>
           #2      AndroidGradleBuilder.buildApk (package:flutter_tools/src/android/gradle.dart:184:5)
           <asynchronous suspension>
           #3      BuildApkCommand.runCommand (package:flutter_tools/src/commands/build_apk.dart:109:5)
           <asynchronous suspension>
           #4      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:1209:27)
           <asynchronous suspension>
```

**Comment by christopherfujino** [comment]
> @christopherfujino I don't think this failure is related to the change:

Retried it

File: packages/flutter_tools/test/integration.shard/debug_adapter/flutter_adapter_test.dart [hunk 1]
```diff
@@ -99,6 +99,13 @@ void main() {
         '',
         startsWith('Exited'),
       ]);
+
+      // If we're running with an out-of-process debug adapter, ensure that its
+      // own process shuts down after we terminated.
+      final DapTestServer server = dap.server;
+      if (server is OutOfProcessDapTestServer) {
```
Comment (christopherfujino) [inline packages/flutter_tools/test/integration.shard/debug_adapter/flutter_adapter_test.dart:106]: Will the concrete class of `server` sometimes be something else when this test is run?

Comment (DanTup) [reply packages/flutter_tools/test/integration.shard/debug_adapter/flutter_adapter_test.dart:106]: You can run the tests with an env variable to un them in-process which is useful for debugging (it's faster, you can step through the code, and you don't need to rebuild flutter_tools):

https://github.com/flutter/flutter/blob/78d924af1d1a8841e693892af73d05686c2f7cb7/packages/flutter_tools/test/integration.shard/debug_adapter/test_support.dart#L25

However in that case, there is no `process` (it only exists on `OutOfProcessDapTestServer`).

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**